bugsnag-plugin-android-ndk/src/main/assets/include/event.h

```diff
@@ -0,0 +1,287 @@
+#ifndef BUGSNAG_ANDROID_NDK_EVENT_API_H
+#define BUGSNAG_ANDROID_NDK_EVENT_API_H
+
+#include <stdbool.h>
+#include <sys/types.h>
+#ifndef BUGSNAG_METADATA_MAX
+/**
+ * Maximum number of values stored in metadata. Configures a default if not
+ * defined.
+ */
+#define BUGSNAG_METADATA_MAX 128
+#endif
+#ifndef BUGSNAG_FRAMES_MAX
+/**
+ *  Number of frames in a stacktrace. Configures a default if not defined.
+ */
+#define BUGSNAG_FRAMES_MAX 192
+#endif
+#ifndef BUGSNAG_CRUMBS_MAX
+/**
+ *  Max number of breadcrumbs in an event. Configures a default if not defined.
+ */
+#define BUGSNAG_CRUMBS_MAX 30
+#endif
+#ifndef BUGSNAG_DEFAULT_EX_TYPE
+/**
+ * Type assigned to exceptions. Configures a default if not defined.
+ */
+#define BUGSNAG_DEFAULT_EX_TYPE "c"
+#endif
+/**
+ * Version of the bugsnag_event struct. Serialized to report header.
+ */
+#define BUGSNAG_EVENT_VERSION 3
+
+#define BUGSNAG_USER_INFO_LEN 64
+
+typedef enum {
+  /** An unhandled exception */
+  BSG_SEVERITY_ERR,
+  /** A handled exception */
+  BSG_SEVERITY_WARN,
+  /** Custom, notable error messages */
+  BSG_SEVERITY_INFO,
+} bsg_severity_t;
+
+typedef enum {
+  /**
+   * Any breadcrumb recorded via \link bugsnag_leave_breadcrumb
+   */
+  BSG_CRUMB_MANUAL,
+  /**
+   * Call to bugsnag_notify*()
+   * (Internal use only)
+   */
+  BSG_CRUMB_ERROR,
+  /**
+   * Log message
+   */
+  BSG_CRUMB_LOG,
+  /**
+   *  Navigation action, such as an overall change in the application view
+   *  context
+   */
+  BSG_CRUMB_NAVIGATION,
+  /**
+   *  Background process, such performing a database query
+   */
+  BSG_CRUMB_PROCESS,
+  /**
+   *  Network request
+   */
+  BSG_CRUMB_REQUEST,
+  /**
+   *  Change in application state
+   */
+  BSG_CRUMB_STATE,
+  /**
+   *  User event, such as authentication or control events
+   */
+  BSG_CRUMB_USER,
+} bsg_breadcrumb_t;
+
+typedef struct {
```

Comment:
The header in `assets/include/event.h` is the public header shown to users who use the Native API. It should not include private properties like the internal representations of stackframe and the `#define` values for stacktrace length, etc, as these aren't public APIs that we want to commit to forever, and we should be able to migrate those values internally without breaking people's code. 

The only public components of event are the severity enum and breadcrumb types, until we design and deploy an API which requires more, (e.g. for callbacks) which will need some design anyway to ensure that the API is clear.

Reply:
_summary of in-person discussion_

There should be a more structured/user-friendly interface for the event prior to implementing `on_error`. In particular, the following points will be addressed as a part of a subsequent document/changeset:

* Having a way to safely set and remove metadata/app/device properties
* Reviewing the interface to ensure it is hardened for arbitrary input as currently the functions were designed for internal use
* Protecting report components which could be incorrectly altered and break reporting (for example, metadata storage and the breadcrumb circular buffer could be broken entirely by changing the index properties incorrectly)

Reply:
I've updated `assets/include/event.h` to only include the severity/breadcrumb type structs, and moved the other structs back to an internal header.

The remaining work on clariying the event API for on_error callbacks will take place in separate PRs.